atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -281,7 +281,8 @@ private synchronized void refreshPool() {
     }
 
     private void addPool(InetSocketAddress server) {
-        addPool(server, new CassandraClientPoolingContainer(server, config));
+        int currentPoolNumber = currentPools.size() + 1;
```

Comment:
I think there are a few problems with this approach although I may be misunderstanding something.

1) `currentPools` can be modified while we are looping through and adding Pools.  This could cause metrics to be re-registered (causing exceptions).  Might be better to maintain a separate counter that is only increased ( `AtomicInteger.incrementAndGet` ) to avoid duplicate metrics.

2) Each product node is going to be registering pool1.metric, pool2.metric but there is no ordering guarantee (or even that they see the same config at start up) so product node1's pool1 metrics will be for a different C* host to node2's pool1 metrics.  There's less we can do about this, perhaps sorting the host names before registering them would help.

3) When a product node is bounced it may register the pools in a different order so graphs may jump, this is probably ok as long as we are aware of it.  Sorting again helps try to keep the order the same.

Reply:
1. I dont think `currentPools` can be modified while we are looping through and adding Pools. The `refreshDaemon` which can modify the pools is scheduled after all the pools are created and metrics are registered.
2. I think internal deployablity tool ensures that same order is discovered across nodes, but agree sorting is a safer option here.
2. Yes, going with sorting the hostnames before adding pools.

Reply:
Note that `config.servers()` is of type `Set<InetSocketAddress>` so it should already be sorted :).

Reply:
`Set` isn't guaranteed to be sorted which is why I wanted to sort them.

Reply:
Yes, you are right. I confused it with `SortedSet`.

Reply:
1. Edit: We can add and remove pools via the refreshDaemon. This is likely to cause issues as:

- We can conflict on numbers when re-adding the hosts.
- We will remove all metrics (for all hosts) from the registry whenever we remove a pool.

 I think a good solution here is to number the servers 1,2,3 and ensure that there is only one number per C* node ever (when registering/deregistering).

Reply:
1. Update: Dived deeper into the metric registration. We have a metricManager per pool and only the pool metrics are deregistered by the manager on removal. We are safe here.. We are basically telling each pool that their prefix is "pool1", "pool2", "pool3", each pool registers the metrics with that prefix and remembers the set of metrics, when we remove a pool and deregister metrics it simply removed the metrics it knows about. So, we are safe here as currentPools will not be modified while we are looping through and adding Pools.

Reply:
so if we register A=pool1, B=pool2, C=pool3, then deregister B, then add D.  What pool does this get?  Seems there no way the system knows that 1 and 3 are in use but 2 isn't.

Think we should either move to a single number that is always incremented (maximum cardinality is the total number of different host addresses used by a product until we can use tags which doesn't seem too high), or write some tests around these cases to prove that they are handled correctly.